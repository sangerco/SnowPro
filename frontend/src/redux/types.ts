export interface UserData {
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    avatar: string;
    bio: string;
    videos: string[];
    photos: string[];
    favMountains: string[];
  }


export const FETCH_USER_DATA_REQUEST = 'FETCH_USER_DATA_REQUEST';
export const FETCH_USER_DATA_SUCCESS = 'FETCH_USER_DATA_SUCCESS';
export const FETCH_USER_DATA_FAILURE = 'FETCH_USER_DATA_FAILURE';


export interface FetchUserDataRequestAction {
    type: typeof FETCH_USER_DATA_REQUEST
}

export interface FetchUserDataSuccessAction {
    type: typeof FETCH_USER_DATA_SUCCESS;
    payload: UserData;
}

export interface FetchUserDataFailureAction {
    type: typeof FETCH_USER_DATA_FAILURE;
    payload: string;
}

export type UserActionTypes = 
    FetchUserDataRequestAction
    | FetchUserDataSuccessAction
    | FetchUserDataFailureAction;