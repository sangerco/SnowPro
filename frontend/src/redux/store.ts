import { Action, compose } from "redux";
import { ThunkAction, configureStore } from "@reduxjs/toolkit";
import { logger } from "redux-logger";
import rootReducer from "./reducers/rootReducer";

const composeEnhancers =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(logger, composeEnhancers),
});
