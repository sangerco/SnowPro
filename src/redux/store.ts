import { configureStore, combineReducers } from "@reduxjs/toolkit";
import userReducer from "./slices/userSlice";
import authReducer from "./slices/authSlice";
import messageReducer from "./slices/messageSlice";
import messageReplyReducer from "./slices/messageReplySlice";
import reviewReducer from "./slices/reviewSlice";
import reviewReplyReducer from "./slices/reviewReplySlice";
import tagReducer from "./slices/tagSlice";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";

const persistConfig = {
  key: "root",
  storage,
  blacklist: ["messages", "messageReplies", "reviews", "reviewReplies"],
};

const rootReducer = combineReducers({
  users: userReducer,
  auth: authReducer,
  messages: messageReducer,
  messageReplies: messageReplyReducer,
  reviews: reviewReducer,
  reviewReplies: reviewReplyReducer,
  tags: tagReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
