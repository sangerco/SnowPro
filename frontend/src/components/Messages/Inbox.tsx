import React, {useEffect} from 'react';
import { connect } from 'react-redux';
import { useParams } from 'react-router';
import { RootState } from '../../oldRedux/store';
import { fetchAllUsersMessagesData, markMessageRead, markMessageUnread } from '../../oldRedux/actions/messageActions';
import { UserWithMessages } from '../../oldRedux/types/messageTypes';
import { Container, Header, Segment, List, Button } from 'semantic-ui-react';
import MessageInboxItem from './MessageInboxItem';

interface InboxProps {
    messages: UserWithMessages[];
    fetchAllUsersMessagesData: (username: string) => void;
    markMessageRead: (id: string) => void;
    markMessageUnread: (id: string) => void;
};

const Inbox: React.FC<InboxProps> = ({ messages, fetchAllUsersMessagesData, markMessageRead, markMessageUnread }) => {
    const { username } = useParams();
    
    useEffect(() => {
        fetchAllUsersMessagesData(username as string);
        }, [username, fetchAllUsersMessagesData]
    );

    const handleMarkAsRead = (id: string) => {
        markMessageRead(id);
    };

    const handleMarkAsUnread = (id: string) => {
        markMessageUnread(id);
    };

    return (
        <Container fluid>
            <Header as='h2'>
                Messages
            </Header>
            <Segment>
                <List divided relaxed>
                    {messages.map((message) => (
                        <List.Item key={message.id}>
                            <List.Content floated='right'>
                                {!message.isRead && 
                                    (<Button onClick={(() => handleMarkAsRead(message.id))} color='blue' size='mini'>
                                        Mark As Read?</Button>)}
                                {message.isRead && 
                                    (<Button onClick={(() => handleMarkAsUnread(message.id))} color='blue' size='mini'>
                                        Mark As Not Read?</Button>)}
                            </List.Content>
                            <MessageInboxItem message={message} />
                        </List.Item>
                    ))}
                </List>
            </Segment>
        </Container>
    )
}

const mapStateToProps = (state: RootState) => ({
    messages: state.message.data,
    error: state.message.error
})

const mapDispatchToProps = {
    fetchAllUsersMessagesData,
    markMessageRead,
    markMessageUnread
}

export default connect(mapStateToProps, mapDispatchToProps)(Inbox)