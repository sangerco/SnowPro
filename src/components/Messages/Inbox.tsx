import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { RootState, AppDispatch } from "../../redux/store";
import { useDispatch, useSelector } from "react-redux";
import {
  fetchUserMessages,
  fetchSentMessages,
  deleteMessage,
  markMessageAsRead,
  markMessageAsUnread,
  MessageData,
} from "../../redux/slices/messageSlice";
import {
  Button,
  Container,
  Dimmer,
  Header,
  List,
  Loader,
  Segment,
} from "semantic-ui-react";

interface InboxProps {
  username: string;
}

const Inbox: React.FC<InboxProps> = ({ username }) => {
  const messages = useSelector((state: RootState) => state.messages);
  const dispatch = useDispatch<AppDispatch>();
  const [showDeleteModal, setShowDeleteModal] = useState(false);

  useEffect(() => {
    dispatch(fetchUserMessages(username));
  }, [dispatch, username]);

  useEffect(() => {
    dispatch(fetchSentMessages(username));
  }, [dispatch, username]);

  const handleMarkAsRead = (id: string) => {
    dispatch(markMessageAsRead(id));
  };

  const handleMarkAsUnread = (id: string) => {
    dispatch(markMessageAsUnread(id));
  };

  const formatDate = (date: Date) => {
    const newDate = new Date(date);
    const options: Intl.DateTimeFormatOptions = {
      year: "numeric",
      month: "long",
      day: "numeric",
    };
    return newDate.toLocaleDateString("en-US", options);
  };

  if (messages.loading) {
    return (
      <Dimmer active>
        <Loader>Loading...</Loader>
      </Dimmer>
    );
  }

  if (messages.error) {
    return <div>Error loading messages! Error: {`${messages.error}`}</div>;
  }

  if (messages.messages) {
    return (
      <Container fluid>
        <Header as="h2">Messages</Header>
        <Segment>
          <Button floated="right" as={Link} to={`/messages/create-message`}>
            New Message
          </Button>
        </Segment>
        <Segment>
          <List divided relaxed>
            {messages.messages.map((message: MessageData) => (
              <List.Item
                key={message.id}
                as={Link}
                to={`/messages/${message.id}`}>
                <List.Content floated="left">{message.subject}</List.Content>
                <List.Content>
                  {message.senderFirstName} {message.senderLastName}
                </List.Content>
                <List.Content floated="right">
                  {formatDate(message.createdAt)}
                </List.Content>
              </List.Item>
            ))}
          </List>
        </Segment>
      </Container>
    );
  }

  return null;
};

export default Inbox;
