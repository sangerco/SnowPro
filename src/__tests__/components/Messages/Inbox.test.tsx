/* eslint-disable testing-library/no-node-access */
import React from "react";
import Inbox from "../../../components/Messages/Inbox";
import { render, fireEvent, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../../redux/store";
import { BrowserRouter } from "react-router-dom";

describe("test Inbox", () => {
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Inbox username={"test1"} />
        </BrowserRouter>
      </Provider>
    );
  });

  it("matches a snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Inbox username={"test1"} />
        </BrowserRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
