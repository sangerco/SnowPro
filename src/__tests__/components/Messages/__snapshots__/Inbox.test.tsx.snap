// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Inbox matches a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui active transition visible dimmer"
    style="display: flex;"
  >
    <div
      class="content"
    >
      <div
        class="ui text loader"
      >
        Loading...
      </div>
    </div>
  </div>
</DocumentFragment>
`;
